File: backend_python/app/repositories/team_repository.py
"""
Team repository - mirrors ITeamRepository
"""
from google.cloud import firestore
from typing import Optional
import logging

logger = logging.getLogger(__name__)


class TeamRepository:
    """
    Repository for team operations.
    Mirrors the .NET TeamRepository.
    """
    
    COLLECTION = "teams"
    
    def __init__(self, db: firestore.Client):
        self.db = db
    
    def _convert_to_dict(self, team: dict) -> dict:
        """Convert team entity to Firestore dictionary"""
        data = {
            "tournamentId": team["tournamentId"],
            "name": team["name"],
            "group": team["group"],
            "captainName": team["captainName"],
            "captainUid": team["captainUid"],
            "logoUrl": team.get("logoUrl", ""),
            "players": [
                {"name": p["name"], "role": p["role"]} 
                for p in team.get("players", [])
            ]
        }
        
        if team.get("substitute"):
            data["substitute"] = {
                "name": team["substitute"]["name"],
                "role": team["substitute"]["role"]
            }
        
        return data
    
    def _convert_from_dict(self, doc_id: str, data: dict) -> dict:
        """Convert Firestore document to team entity"""
        players = []
        if "players" in data and data["players"]:
            players = [
                {"name": p.get("name", ""), "role": p.get("role", "")}
                for p in data["players"]
            ]
        
        substitute = None
        if "substitute" in data and data["substitute"]:
            sub_data = data["substitute"]
            substitute = {
                "name": sub_data.get("name", ""),
                "role": sub_data.get("role", "")
            }
        
        return {
            "id": doc_id,
            "tournamentId": data.get("tournamentId", ""),
            "name": data.get("name", ""),
            "group": data.get("group", ""),
            "captainName": data.get("captainName", ""),
            "captainUid": data.get("captainUid", ""),
            "logoUrl": data.get("logoUrl", ""),
            "players": players,
            "substitute": substitute
        }
    
    async def create(self, team: dict) -> str:
        """Create a team - mirrors CreateAsync"""
        doc_data = self._convert_to_dict(team)
        doc_ref = self.db.collection(self.COLLECTION).document()
        doc_ref.set(doc_data)
        return doc_ref.id
    
    async def get_by_id(self, team_id: str) -> Optional[dict]:
        """Get team by ID - mirrors GetByIdAsync"""
        doc_ref = self.db.collection(self.COLLECTION).document(team_id)
        doc_snapshot = doc_ref.get()
        
        if not doc_snapshot.exists:
            return None
        
        return self._convert_from_dict(doc_snapshot.id, doc_snapshot.to_dict())
    
    async def get_by_tournament_id(self, tournament_id: str) -> list[dict]:
        """Get teams by tournament ID - mirrors GetByTournamentIdAsync"""
        docs = self.db.collection(self.COLLECTION).where("tournamentId", "==", tournament_id).stream()
        results = []
        for doc in docs:
            results.append(self._convert_from_dict(doc.id, doc.to_dict()))
        return results
    
    async def update(self, team: dict) -> None:
        """Update team - mirrors UpdateAsync"""
        team_id = team["id"]
        doc_data = self._convert_to_dict(team)
        doc_ref = self.db.collection(self.COLLECTION).document(team_id)
        doc_ref.set(doc_data, merge=True)
    
    async def delete(self, team_id: str) -> None:
        """Delete team - mirrors DeleteAsync"""
        doc_ref = self.db.collection(self.COLLECTION).document(team_id)
        doc_ref.delete()

File: backend/CricketTournament.Infrastructure/Repositories/TeamRepository.cs
using CricketTournament.Domain.Entities;
using CricketTournament.Domain.Interfaces;
using System.Collections;

namespace CricketTournament.Infrastructure.Repositories;

public class TeamRepository : ITeamRepository
{
    private readonly IFirestoreRepository _firestoreRepository;
    private const string Collection = "teams";

    public TeamRepository(IFirestoreRepository firestoreRepository)
    {
        _firestoreRepository = firestoreRepository;
    }

    public async Task<string> CreateAsync(Team team)
    {
        var data = ConvertToDictionary(team);
        return await _firestoreRepository.CreateDocumentAsync(Collection, data);
    }

    public async Task<Team?> GetByIdAsync(string id)
    {
        var data = await _firestoreRepository.GetDocumentAsync<Dictionary<string, object>>(Collection, id);
        return data == null ? null : ConvertFromDictionary(id, data);
    }

    public async Task<List<Team>> GetByTournamentIdAsync(string tournamentId)
    {
        var documents = await _firestoreRepository.GetDocumentsAsync<Dictionary<string, object>>(Collection, "tournamentId", tournamentId);
        return documents.Select(doc => ConvertFromDictionary(doc["id"]?.ToString() ?? string.Empty, doc)).ToList();
    }

    public async Task UpdateAsync(Team team)
    {
        var data = ConvertToDictionary(team);
        await _firestoreRepository.UpdateDocumentAsync(Collection, team.Id, data);
    }

    public async Task DeleteAsync(string id)
    {
        await _firestoreRepository.DeleteDocumentAsync(Collection, id);
    }

    private static Dictionary<string, object> ConvertToDictionary(Team team)
    {
        var data = new Dictionary<string, object>
        {
            { "tournamentId", team.TournamentId },
            { "name", team.Name },
            { "group", team.Group },
            { "captainName", team.CaptainName },
            { "captainUid", team.CaptainUid },
            { "logoUrl", team.LogoUrl },
            { "players", team.Players.Select(p => new Dictionary<string, object> { { "name", p.Name }, { "role", p.Role } }).ToList() }
        };

        if (team.Substitute != null)
        {
            data["substitute"] = new Dictionary<string, object> { { "name", team.Substitute.Name }, { "role", team.Substitute.Role } };
        }

        return data;
    }

    private static Team ConvertFromDictionary(string id, Dictionary<string, object> data)
    {
        var players = new List<Player>();
        if (data.ContainsKey("players") && data["players"] is List<object> playersList)
        {
            players = playersList.Select(p =>
            {
                var playerDict = (Dictionary<string, object>)p;
                return new Player
                {
                    Name = playerDict["name"]?.ToString() ?? string.Empty,
                    Role = playerDict["role"]?.ToString() ?? string.Empty
                };
            }).ToList();
        }

        Player? substitute = null;
        if (data.ContainsKey("substitute") && data["substitute"] is Dictionary<string, object> subDict)
        {
            substitute = new Player
            {
                Name = subDict["name"]?.ToString() ?? string.Empty,
                Role = subDict["role"]?.ToString() ?? string.Empty
            };
        }

        return new Team
        {
            Id = id,
            TournamentId = data["tournamentId"]?.ToString() ?? string.Empty,
            Name = data["name"]?.ToString() ?? string.Empty,
            Group = data["group"]?.ToString() ?? string.Empty,
            CaptainName = data["captainName"]?.ToString() ?? string.Empty,
            CaptainUid = data["captainUid"]?.ToString() ?? string.Empty,
            LogoUrl = data["logoUrl"]?.ToString() ?? string.Empty,
            Players = players,
            Substitute = substitute
        };
    }
}

